Binary files original.tuxman-0.7/bin/tuxman and tuxman-0.7/bin/tuxman differ
diff -Nur original.tuxman-0.7/Makefile tuxman-0.7/Makefile
--- original.tuxman-0.7/Makefile	2004-01-01 12:41:17.000000000 +0100
+++ tuxman-0.7/Makefile	2004-01-11 18:11:32.000000000 +0100
@@ -1,8 +1,8 @@
 all: bin/tuxman
 
 CXX = g++
-LDFLAGS = `sdl-config --libs`
-CXXFLAGS = `sdl-config --cflags`
+LDFLAGS = `sdl-config --libs` -lggzmod
+CXXFLAGS = `sdl-config --cflags` -ggdb3 -Wall
 
 bin/tuxman: obj/main.o \
 	obj/pacman.o \
@@ -20,3 +20,6 @@
 
 clean:
 	rm obj/*
+
+ggzinstall:
+	ggz-config --install --modfile=module.dsc
diff -Nur original.tuxman-0.7/module.dsc tuxman-0.7/module.dsc
--- original.tuxman-0.7/module.dsc	1970-01-01 01:00:00.000000000 +0100
+++ tuxman-0.7/module.dsc	2004-01-01 12:41:18.000000000 +0100
@@ -0,0 +1,11 @@
+[ModuleInfo]
+Game = Tuxman
+Author = Robert Strobl
+CommandLine = ./bin/tuxman --ggz
+Frontend = sdl
+Homepage = http://www.ftpnet.org/
+Name = Tuxman
+ProtocolEngine = Tuxman
+ProtocolVersion = 1
+Version = 0.7
+
Binary files original.tuxman-0.7/obj/control.o and tuxman-0.7/obj/control.o differ
Binary files original.tuxman-0.7/obj/main.o and tuxman-0.7/obj/main.o differ
Binary files original.tuxman-0.7/obj/pacman.o and tuxman-0.7/obj/pacman.o differ
Binary files original.tuxman-0.7/obj/sound.o and tuxman-0.7/obj/sound.o differ
Binary files original.tuxman-0.7/obj/tile.o and tuxman-0.7/obj/tile.o differ
Binary files original.tuxman-0.7/screen.bmp and tuxman-0.7/screen.bmp differ
diff -Nur original.tuxman-0.7/src/control.cpp tuxman-0.7/src/control.cpp
--- original.tuxman-0.7/src/control.cpp	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/control.cpp	2004-01-11 18:49:34.000000000 +0100
@@ -1,4 +1,16 @@
 #include "control.h"
+#include <iostream>
+
+void control::networkMove(pacman &pac, int direction)
+{
+#ifdef GGZ_SUPPORT
+	  if(pac.ggzmode) {
+		std::cout << "event: move " << pac.ggzfd << std::endl;
+		ggz_write_int(pac.ggzfd, PACNET::move);
+		ggz_write_int(pac.ggzfd, direction);
+	  }
+#endif
+}
 
 void control::checkEvent(pacman &pac, tile &my_tile) {
   SDL_Event ev;
@@ -11,20 +23,31 @@
       case SDLK_ESCAPE:
 	  pac.quit();
       case SDLK_LEFT:
-	if (pac.pause != PAUSE)
-	  my_tile.my_pac.state=LEFT;
+	if (pac.pause != PAUSE) {
+	  if(!pac.ggzmode) my_tile.my_pac.state=LEFT;
+	  networkMove(pac, PACNET::left);
+	}
         break;
       case SDLK_RIGHT:
 	if (pac.pause != PAUSE)
-	  my_tile.my_pac.state=RIGHT;
+	{
+	  if(!pac.ggzmode) my_tile.my_pac.state=RIGHT;
+	  networkMove(pac, PACNET::right);
+	}
         break;
       case SDLK_UP:
 	if (pac.pause != PAUSE)
-	  my_tile.my_pac.state=UP;
+	{
+	  if(!pac.ggzmode) my_tile.my_pac.state=UP;
+	  networkMove(pac, PACNET::up);
+	}
         break;
       case SDLK_DOWN:
 	if (pac.pause != PAUSE)
-	  my_tile.my_pac.state=DOWN;
+	{
+	  if(!pac.ggzmode) my_tile.my_pac.state=DOWN;
+	  networkMove(pac, PACNET::down);
+	}
         break;
       case SDLK_PLUS:
 	pac.GameSpeed-=1;	
diff -Nur original.tuxman-0.7/src/control.h tuxman-0.7/src/control.h
--- original.tuxman-0.7/src/control.h	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/control.h	2004-01-01 12:41:18.000000000 +0100
@@ -8,6 +8,38 @@
 
 #define SCROLL_SIZE 6
 
+namespace PACNET
+{
+	enum Opcodes
+	{
+		list = 100,
+		name = 101,
+		monster = 102,
+		event = 103,
+		pacman = 104,
+		selected = 120,
+		move = 121
+	};
+	enum Events
+	{
+		error = 1,
+		start = 2
+	};
+	enum Moves
+	{
+		up = 1,
+		left = 2,
+		down = 3,
+		right = 4
+	};
+	enum Errors
+	{
+		undef = 1,
+		net = 2,
+		map = 3
+	};
+};
+
 class control {
 
  public:
@@ -22,6 +54,9 @@
   void checkMonsterMoving(tile &my_tile);
   void checkPacmansDeath(tile &my_tile, pacman &pac);
   void checkScroll(tile &my_tile, pacman &pac);
+
+ private:
+  void networkMove(pacman &pac, int direction);
 };
 
 #endif
diff -Nur original.tuxman-0.7/src/main.cpp tuxman-0.7/src/main.cpp
--- original.tuxman-0.7/src/main.cpp	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/main.cpp	2004-01-11 19:15:38.000000000 +0100
@@ -2,6 +2,159 @@
 #include "control.h"
 #include "tile.h"
 #include "sound.h"
+#ifdef GGZ_SUPPORT
+#include <ggzmod.h>
+#include <iostream>
+#endif
+
+#ifdef GGZ_SUPPORT
+pacman *pacptr = NULL;
+control *ctrlptr = NULL;
+tile *tileptr = NULL;
+static GGZMod *mymod = NULL;
+
+static void pac_ggzmod(GGZMod *mod, GGZModEvent e, void *data)
+{
+	pacptr->ggzfd = *(int*)data;
+	ggzmod_set_state(mod, GGZMOD_STATE_PLAYING);
+}
+
+static void pac_ggz()
+{
+	ggzmod_dispatch(mymod);
+}
+
+static void pac_input()
+{
+	int opcode, fd, arg;
+	int x, y, direction, count, pos;
+	char *level;
+	bool hasinput;
+
+	fd = pacptr->ggzfd;
+
+	hasinput = true;
+	while(hasinput)
+	{
+		ggz_read_int(fd, &opcode);
+
+		std::cout << "pacnet: input opcode=" << opcode << std::endl << std::flush;
+
+		switch(opcode)
+		{
+			case PACNET::list:
+				std::cout << "pacnet: got level list, sending dummy" << std::endl;
+				ggz_read_int(fd, &count);
+				for(int i = 0; i < count; i++)
+				{
+					ggz_read_string_alloc(fd, &level);
+					std::cout << " * level " << level << std::endl;
+					ggz_free(level);
+				}
+				ggz_write_int(fd, PACNET::selected);
+				ggz_write_string(fd, "dummy.level");
+				break;
+			case PACNET::name:
+				/* ... ? */
+				break;
+			case PACNET::monster:
+				ggz_read_int(fd, &x);
+				ggz_read_int(fd, &y);
+				ggz_read_int(fd, &direction);
+				std::cout << "pacnet: move monster" << std::endl;
+				pos = tileptr->width * y + x;
+				std::cout << "would move monster at " << pos << std::endl;
+				if(direction == PACNET::up) tileptr->my_tile[pos].row -= TILE_SIZE;
+				else if(direction == PACNET::down) tileptr->my_tile[pos].row += TILE_SIZE;
+				else if(direction == PACNET::left) tileptr->my_tile[pos].col -= TILE_SIZE;
+				else if(direction == PACNET::right) tileptr->my_tile[pos].col += TILE_SIZE;
+				break;
+			case PACNET::event:
+				ggz_read_int(fd, &arg);
+				if(arg == PACNET::start)
+				{
+					std::cout << "pacnet: start event" << std::endl;
+				}
+				else if(arg == PACNET::error)
+				{
+					std::cout << "pacnet: error event" << std::endl;
+				}
+				else;
+				break;
+			case PACNET::pacman:
+				ggz_read_int(fd, &x);
+				ggz_read_int(fd, &y);
+				ggz_read_int(fd, &direction);
+				std::cout << "pacnet: move pacman" << std::endl;
+				if(direction == PACNET::up) {
+					tileptr->my_pac.state = UP;
+				} else if(direction == PACNET::down) {
+					tileptr->my_pac.state = DOWN;
+				} else if(direction == PACNET::left) {
+					tileptr->my_pac.state = LEFT;
+				} else if(direction == PACNET::right) {
+					tileptr->my_pac.state = RIGHT;
+				}
+				break;
+			default:
+				std::cout << "pacnet: error" << std::endl;
+				// error
+		}
+
+		fd_set tmp;
+		struct timeval tv;
+		tv.tv_sec = 0;
+		tv.tv_usec = 0;
+
+		FD_ZERO(&tmp);
+		FD_SET(fd, &tmp);
+		select(fd + 1, &tmp, NULL, NULL, &tv);
+		if(!FD_ISSET(fd, &tmp)) hasinput = false;
+	}
+}
+
+/* taken from sdl-games/ttt3d */
+static void pac_network()
+{
+	int maxfd;
+	fd_set set;
+	struct timeval tv;
+	int ret;
+	int serverfd;
+	int fdes;
+
+	fdes = pacptr->ggzfd;
+
+	FD_ZERO(&set);
+
+	serverfd = ggzmod_get_fd(mymod);
+
+	FD_SET(serverfd, &set);
+	maxfd = serverfd;
+	if(fdes >= 0)
+	{
+		FD_SET(fdes, &set);
+		if(fdes > maxfd) maxfd = fdes;
+	}
+
+	tv.tv_sec = 0;
+	tv.tv_usec = 0;
+
+	ret = select(maxfd + 1, &set, NULL, NULL, &tv);
+
+	if(ret)
+	{
+		if(fdes >= 0)
+		{
+			if(FD_ISSET(fdes, &set))
+			{
+				pac_input();
+			}
+		}
+		if(FD_ISSET(serverfd, &set)) pac_ggz();
+	}
+}
+#endif
 
 int main(int argc, char *argv[]) {
   int food, i;
@@ -12,6 +165,20 @@
 
   pac.readConfig("tuxman.conf",pac);  
 
+#ifdef GGZ_SUPPORT
+  if((argc == 2) && (!strcmp(argv[1], "--ggz"))) {
+	pac.ggzmode = 1;
+	pac.ggzfd = -1;
+	pacptr = &pac;
+	ctrlptr = &ctrl;
+	tileptr = &my_tile;
+  } else {
+	pac.ggzmode = 0;
+  }
+#else
+  pac.ggzmode = 0;
+#endif
+
   srand(time(0));
   
   pac.initSDL();
@@ -25,12 +192,34 @@
   pac.ac_level=1;
   my_tile.my_pac.score=0;
   my_tile.my_pac.lives = 10;
- 
+
+#ifdef GGZ_SUPPORT
+  if(pac.ggzmode) {
+	std::cout << "start GGZ connection" << std::endl;
+    pac.ggzmod = ggzmod_new(GGZMOD_GAME);
+	mymod = pac.ggzmod;
+	ggzmod_set_handler(pac.ggzmod, GGZMOD_EVENT_SERVER, &pac_ggzmod);
+    ggzmod_connect(pac.ggzmod);
+  }
+#endif
+
+#ifdef GGZ_SUPPORT
+	if(pac.ggzmode) {
+		while(ggzmod_get_state(pac.ggzmod) != GGZMOD_STATE_PLAYING)
+			pac_network();
+		std::cout << "got GGZ connection" << std::endl;
+    }
+#endif
+
   while(1) {
     pac.nextLevel();
-    if (argc>=2)
+    if ((!pac.ggzmode) && (argc>=2)) {
       sprintf(pac.level,"%s",argv[1]);
- 
+	}
+	if(pac.ggzmode) {
+		sprintf(pac.level, "levels/level9.pac");
+	}
+
     my_tile.initTiles(pac);
     playSound("share/hunger.wav");
     SDL_PauseAudio(0);
@@ -46,6 +235,9 @@
 	if (food==0)
 	  break;
 
+#ifdef GGZ_SUPPORT
+	if(pac.ggzmode) pac_network();
+#endif
 	ctrl.checkPacmansDeath(my_tile, pac);
 	ctrl.checkScroll(my_tile,pac);	
       }
@@ -58,11 +250,13 @@
       }
 
       if (pac.pause != PAUSE) {
-	
+
 	ctrl.checkState(my_tile);
 	ctrl.checkFoodEating(my_tile, pac);
 	ctrl.checkWallCollision(my_tile);
-	ctrl.checkMonsterMoving(my_tile);
+	if (!pac.ggzmode) {
+		ctrl.checkMonsterMoving(my_tile);
+	}
 
 	my_tile.drawTiles(pac);
 	my_tile.showInformation(pac);
@@ -74,7 +268,15 @@
     playSound("share/end.wav");
     SDL_PauseAudio(0);
     sleep(2);
-  } 
+  }
+
+#ifdef GGZ_SUPPORT
+  if(pac.ggzmode) {
+	ggzmod_disconnect(pac.ggzmod);
+	ggzmod_free(pac.ggzmod);
+  }
+#endif
+
   pac.quit();
 }
 
diff -Nur original.tuxman-0.7/src/main.h tuxman-0.7/src/main.h
--- original.tuxman-0.7/src/main.h	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/main.h	2004-01-01 12:41:18.000000000 +0100
@@ -11,6 +11,8 @@
 #include <string.h>
 #include <ctype.h>
 
+#define GGZ_SUPPORT         1
+
 #define TILE_SIZE           32
 
 #define ICON                ""
diff -Nur original.tuxman-0.7/src/pacman.h tuxman-0.7/src/pacman.h
--- original.tuxman-0.7/src/pacman.h	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/pacman.h	2004-01-01 12:41:18.000000000 +0100
@@ -2,6 +2,9 @@
 #define _pacman_h_
 
 #include "main.h"
+#ifdef GGZ_SUPPORT
+#include <ggzmod.h>
+#endif
 
 class pacman {
 public:
@@ -37,6 +40,12 @@
   int frame_time;
   int frame_last;
 
+  int ggzmode;
+#ifdef GGZ_SUPPORT
+  GGZMod *ggzmod;
+  int ggzfd;
+#endif
+
   //Prototypes
   void initSDL();
   void setVideoMode();
diff -Nur original.tuxman-0.7/src/sound.cpp tuxman-0.7/src/sound.cpp
--- original.tuxman-0.7/src/sound.cpp	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/sound.cpp	2004-01-11 15:32:23.000000000 +0100
@@ -39,7 +39,7 @@
   
   if ( SDL_OpenAudio(&music, NULL) < 0 ) {
     fprintf(stderr, "Couldn't open audio: %s\n", SDL_GetError());
-    pac.quit();
+    /*pac.quit();*/
   }
 }
 
diff -Nur original.tuxman-0.7/src/tile.cpp tuxman-0.7/src/tile.cpp
--- original.tuxman-0.7/src/tile.cpp	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/tile.cpp	2004-01-11 15:34:01.000000000 +0100
@@ -22,9 +22,10 @@
   int row = 0;
   int col = 0;
   
-  if(!(tiles=fopen(pac.level,"r+")))
+  if(!(tiles=fopen(pac.level,"r")))
     printf("Error: Couldn't read %s\n",pac.level);
-  
+
+  width = 0;
   while((rd=fgetc(tiles))!= EOF) { 
     if (rd == '#')
       my_tile[i].type = WALL;
@@ -63,6 +64,7 @@
   
     if (rd == '\n') {
       row+=1;
+	  if(col > width) width = col;
       col=(-1);
     }
     col++;
@@ -281,16 +283,22 @@
   
   SDL_BlitSurface(pac.tuxman, &src, pac.screen, &dest); 
   
-  sprintf(font_string,"A GAME BY ROBERT STROBL 2003",my_pac.lives);
+  sprintf(font_string,"A GAME BY ROBERT STROBL 2003");
   drawFont((pac.SCREEN_WIDTH / 2) - (pac.tuxman->w / 2),
 	   (pac.SCREEN_HEIGHT / 2), pac);
 
-  sprintf(font_string,"WWW FTPNET ORG",my_pac.lives);
+  sprintf(font_string,"WWW FTPNET ORG");
   drawFont((pac.SCREEN_WIDTH / 2) - (pac.tuxman->w / 2),
 	   (pac.SCREEN_HEIGHT / 2) + FONT_HEIGHT + 5, pac);
 
+  if(pac.ggzmode) {
+    sprintf(font_string,"RUNNING IN GGZ MODE");
+    drawFont((pac.SCREEN_WIDTH / 2) - (pac.tuxman->w / 2),
+	   (pac.SCREEN_HEIGHT / 2) + FONT_HEIGHT * 2 + 10, pac);
+  }
+
   SDL_UpdateRect(pac.screen, 0, 0, 0, 0);
-  sleep(5);
+  sleep(3);
 
   pac.BGColor = SDL_MapRGB(pac.screen->format, 97, 144, 181);
   SDL_FillRect(pac.screen, 0, pac.BGColor);
diff -Nur original.tuxman-0.7/src/tile.h tuxman-0.7/src/tile.h
--- original.tuxman-0.7/src/tile.h	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/src/tile.h	2004-01-01 12:41:18.000000000 +0100
@@ -15,6 +15,7 @@
  public:
   int monster;
   int tile_max;
+  int width;
   char font_string[256];
 
   struct tileset {
diff -Nur original.tuxman-0.7/tuxman.conf tuxman-0.7/tuxman.conf
--- original.tuxman-0.7/tuxman.conf	2004-01-01 12:41:18.000000000 +0100
+++ tuxman-0.7/tuxman.conf	2004-01-01 12:41:18.000000000 +0100
@@ -2,10 +2,12 @@
 # Please edit your config file!
 
 # screen width is you width in pixels
-ScreenWidth 1024
+ScreenWidth 640
+#ScreenWidth 1024
 
 # screen height is you height in pixels
-ScreenHeight 768
+ScreenHeight 480
+#ScreenHeight 768
 
 # screen depth 8/16/24/32
 ScreenDepth 24
@@ -16,4 +18,4 @@
 # speed of your game in milliseconds
 # max = 0
 # min = not defined ;)
-GameSpeed 0
+GameSpeed 50
